### PR TITLE
Add CI workflow

### PR DESCRIPTION
**Pull request by archanaserver**
This pull request adds a GitHub Actions CI workflow to this project. 

**Changes Made:**
- Created `.github/workflows/ci.yml` to define the CI workflow.
- Updated the gemspec file to specify development and runtime dependencies.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  test:
+    name: Ruby
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        ruby:
+          - 2.7
+          - 3.0
+          - 3.1
```
Comment (ekohl) [inline .github/workflows/ci.yaml:21]: You should always quote the versions. Otherwise it may end up being interpreted as floats. So `3.0` becomes `3` which other tooling can interpret as "the latest Ruby 3".

**Review by ekohl — COMMENTED** [review]
I wrote a quick reusable action. https://github.com/theforeman/actions#gem-test shows a basic version of how to use it. In https://github.com/ekohl/erb2epp/blob/master/.github/workflows/test.yml I just made use of it (which triggered me to write it). You also need to provide an input for the test command.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          - "3.0"
+          - "3.1"
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+      - name: Set up Ruby
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: ${{ matrix.ruby }}
+          bundler-cache: true
```
Comment (ekohl) [inline .github/workflows/ci.yaml:29]: I've implemented both the matrix and code checkout + ruby setup inside the reusable workflow. https://github.com/theforeman/actions#gem-test should be the complete example, except that you should specify the Ruby versions via `with` like you also specified `test`.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (ekohl) [reply .github/workflows/ci.yaml:29]: This is now done and https://github.com/theforeman/actions#gem-test is now the complete example (except for the test command). It automatically detects the compatible Ruby versions.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -4,13 +4,15 @@ on:
...
-      test: 'bundle exec rake test'
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Tests
+        uses: theforeman/actions/.github/workflows/test-gem.yml@v0
+
+      - name: Run tests
+        run: |
+          bundle exec rake test
```
Comment (archanaserver) [inline .github/workflows/ci.yaml:18]: @ekohl can you help me to understand this, what is wrong here if I'm doing it this or like what would be correct way to do it? I tried multiple ways but it kept failing.

Comment (ekohl) [reply .github/workflows/ci.yaml:18]: https://docs.github.com/en/actions/using-workflows/reusing-workflows#passing-inputs-and-secrets-to-a-reusable-workflow documents how passing inputs to reusable workflows works, but I've also just added an example: https://github.com/theforeman/actions#gem-test

Comment (archanaserver) [reply .github/workflows/ci.yaml]: is it checking for ruby 3.2 too? do we want to check this?

Comment (archanaserver) [reply .github/workflows/ci.yaml]: i was looking at this one https://github.com/archanaserver/kafo_parsers/actions/runs/5611152228
but seems like i need update development dependencies versions on gemspec to pass the test.

Comment (archanaserver) [reply .github/workflows/ci.yaml]: @ekohl does this changes https://github.com/theforeman/kafo_parsers/commit/a1a41b07d633a21ae580b772f357636ecad183e6 looks good? I guess this is all steps we need right?

**Review by ekohl — CHANGES_REQUESTED** [review]
If you fix the dependencies to have a version spec it should pass. Some filled in:
```ruby
  spec.add_development_dependency "rake", "~> 13.0"
  spec.add_development_dependency "minitest", "~> 5.0"
  spec.add_development_dependency "simplecov"
  spec.add_development_dependency "ci_reporter_minitest"

  # puppet manifests parsing
  spec.add_dependency 'rdoc', '>= 3.9.0', '< 7'
  spec.add_dependency 'json'
```

**Comment by archanaserver** [comment]
@ekohl can we skip ruby 2.5 and 2.4 version checks here?

File: kafo_parsers.gemspec [hunk 4]
```diff
@@ -16,13 +16,15 @@ Gem::Specification.new do |spec|
   spec.test_files    = spec.files.grep(%r{^(test|spec|features)/})
   spec.require_paths = ["lib"]
 
+  spec.required_ruby_version = '>= 2.7', '< 4'
```
Comment (ekohl) [inline kafo_parsers.gemspec:19]: This is one of those things that I always put in its own commit because it's WAY more important than the other parts. This impacts the usage elsewhere while the other bits less so (though runtime dependencies do matter as well). It's still OK to put in the same PR, but then I'd also change the PR summary to reflect it.

File: kafo_parsers.gemspec [hunk 5]
```diff
@@ -16,13 +16,15 @@ Gem::Specification.new do |spec|
...
+  spec.add_development_dependency "rake", "~> 13.0"
+  spec.add_development_dependency "minitest", "~> 5.0"
+  spec.add_development_dependency "simplecov", "~> 0.21"
+  spec.add_development_dependency "ci_reporter_minitest", "~> 1.0"
 
   # puppet manifests parsing
-  spec.add_dependency 'rdoc', '>= 3.9.0'
-  spec.add_dependency 'json'
+  spec.add_dependency "rdoc", ">= 3.9.0", "< 7"
+  spec.add_dependency "json", "~> 2.6"
```
Comment (ekohl) [inline kafo_parsers.gemspec:29]: json is a bit special. It's actually shipped as part of Ruby and version 2.6 is quite aggressive. Here I'd use `~> 2.0` because Ruby 2.7 bundles json 2.3.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: kafo_parsers.gemspec (hunk 4 above)
Comment (archanaserver) [reply kafo_parsers.gemspec]: @ekohl https://github.com/theforeman/kafo_parsers/pull/52

Thanks for the mention!

**Review by ekohl — COMMENTED** [review]
I've merged #52 so if you rebase this then I expect it to pass. Please squash the first and second commit (those that touch `ci.yaml`) together.

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    branches:
+      - master
+
+jobs:
+  test:
+    name: Tests
+    uses: theforeman/actions/.github/workflows/test-gem.yml@v0
+    with:
+      command: bundle exec rake test
+
```
Comment (ekohl) [inline .github/workflows/ci.yaml:15]: This empty line can be dropped
```suggestion
```

**Review by ekohl — APPROVED** [review]
Small nit, but otherwise :+1:

File: .github/workflows/ci.yaml (hunk 6 above)
Comment (archanaserver) [reply .github/workflows/ci.yaml]: Done!

**Comment by archanaserver** [comment]
@ekohl does this need any other changes, i guess it is ready to be merged?

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+on:
+  pull_request:
+  push:
+
+jobs:
+  test:
+    name: Tests
+    uses: theforeman/actions/.github/workflows/test-gem.yml@v0
+    with:
+      command: bundle exec rake test
```
[the remark this answers is not in the document]
Comment (ekohl) [reply .github/workflows/ci.yaml:14]: You still see the icon that it's not there. I'm not sure which editor you use, but I'd recommend making sure that it enforces them.

**Review by ekohl — APPROVED** [review]
